Added TagHelperResolutionResult Converter

TanayParikh opened the pull request:
Fixes: https://github.com/dotnet/aspnetcore/issues/23597

Essentially instead of `RazorLanguageService.GetTagHelpersAsync` returning a `JObject` to Roslyn, and consequently to `OOPTagHelperResolver` which has to convert it back to `TagHelperResolutionResult`, we communicate directly with the `TagHelperResolutionResult` object. This should significantly reduce the overhead of repeated conversions to/from JObject (and related reflection) and consequently, reduce memory utilization. 


Overall: 61% lower CPU time & 64% lower memory utilization

|                                                                  Method |      Mean |     Error |    StdDev |   Op/s |    Gen 0 |   Gen 1 |   Gen 2 | Allocated |
|------------------------------------------------------------------------ |----------:|----------:|----------:|-------:|---------:|--------:|--------:|----------:|
|       'Razor TagHelperResolutionResult Roundtrip JObject Serialization' | 12.221 ms | 0.1034 ms | 0.0916 ms |  81.83 | 125.0000 | 78.1250 | 31.2500 |  10.57 MB |
| 'Razor TagHelperResolutionResult Roundtrip JsonConverter Serialization' |  4.707 ms | 0.0415 ms | 0.0368 ms | 212.46 |  70.3125 | 46.8750 | 31.2500 |   3.85 MB |

Tried a streaming based approach, however this led to a much simpler solution. Thanks @ntaylormullen!

#### Streaming based solution
For future reference; should we need to explore it again:
1. Razor OOPTagHelperResolver.ResolveTagHelpersOutOfProcessAsync -> passes in a data reader which reads in a stream and converts to the TagHelperResolutionResult

`<RPC>`

2. Roslyn Compilation RemoteEndpoint.InvokeAsync -> Leverages https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/Remote/RemoteHostClient.cs#L80-L84

`<RPC>`

3. ServiceHub.RoslynCodeAnalysisService (Razor Managed) - RazorLanguageService.GetTagHelpersAsync -> implements a new data writer which writes the TagHelperResolutionResult to a stream

`<RPC>`

4. Roslyn (captures response from #3) RemoteEndpoint.InvokeAsync -> Roslyn gets the stream from #3, uses the data reader from #1 to get the TagHelperResolutionResult
https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Remote/ServiceHub/Shared/RemoteEndPoint.cs#L188-L205

`<RPC>`

5. Razor OOPTagHelperResolver.ResolveTagHelpersOutOfProcessAsync gets final response as a TagHelperResolutionResult

TanayParikh commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperResolutionResultJsonConverter.cs:
```diff
@@ -0,0 +1,74 @@
+            }
+            writer.WriteEndArray();
+
+            // NOTE: This doesn't currently support razor diagnostics
```
We return empty for `RazorDiagnostic` regardless
https://github.com/dotnet/aspnetcore-tooling/blob/master/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TagHelperResolver.cs#L54

NTaylorMullen commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperResolutionResultJsonConverter.cs:
```diff
@@ -0,0 +1,74 @@
+
+        public override bool CanConvert(Type objectType)
+        {
+            return typeof(TagHelperResolutionResult).IsAssignableFrom(objectType);
```
Do a strict equals comparison. It's faster. We don't do that in TagHelper land only because there can be mock TagHelpers

TanayParikh commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TagHelperResolutionResult.cs:
```diff
@@ -4,9 +4,12 @@
 
 namespace Microsoft.CodeAnalysis.Razor
 {
+    [JsonConverter(typeof(TagHelperResolutionResultJsonConverter))]
```
https://www.newtonsoft.com/json/help/html/Performance.htm#JsonConverters

NTaylorMullen commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperResolutionResultJsonConverter.cs:
```diff
@@ -0,0 +1,74 @@
+            var descriptors = new List<TagHelperDescriptor>();
+            var descriptorConverter = TagHelperDescriptorJsonConverter.Instance;
+
+            if (reader.TokenType != JsonToken.PropertyName ||
```
This whole if statement is quite confusing. Could you break it into many with comments

TanayParikh commented:
> Curious what the perf differences are in the behavior

61% lower CPU time & 64% lower memory utilization

|                                                                  Method |      Mean |     Error |    StdDev |   Op/s |    Gen 0 |   Gen 1 |   Gen 2 | Allocated |
|------------------------------------------------------------------------ |----------:|----------:|----------:|-------:|---------:|--------:|--------:|----------:|
|       'Razor TagHelperResolutionResult Roundtrip JObject Serialization' | 12.221 ms | 0.1034 ms | 0.0916 ms |  81.83 | 125.0000 | 78.1250 | 31.2500 |  10.57 MB |
| 'Razor TagHelperResolutionResult Roundtrip JsonConverter Serialization' |  4.707 ms | 0.0415 ms | 0.0368 ms | 212.46 |  70.3125 | 46.8750 | 31.2500 |   3.85 MB |

TanayParikh commented:
Added benchmarks and tests, as well as a major refactoring. We had the `RazorDiagnosticsJsonConverter` elsewhere (and duplicated). I consolidated the copies into one place, and added in support for `RazorDiagnostics` for the `TagHelperResolutionResult`.

TanayParikh commented on src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Serialization/TagHelperResolutionResultSerializationTest.cs:
```diff
@@ -0,0 +1,354 @@
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    public class TagHelperResolutionResultSerializationTest
```
Tests are just adapted from `TagHelperDescriptorSerializationTest`s

TanayParikh commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TagHelperResolutionResultComparer.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Microsoft.CodeAnalysis.Razor
+{
+    internal sealed class TagHelperResolutionResultComparer : IEqualityComparer<TagHelperResolutionResult>
```
Needed for testing.

TanayParikh commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs:
```diff
@@ -849,44 +848,5 @@ private static SourceSpan ReadSourceSpan(JsonReader reader)
             return sourceSpan;
         }
-
-        private static void ReadProperties(JsonReader reader, Action<string> onProperty)
```
Moved to `JsonReaderExtensions`

TanayParikh commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/RazorDiagnosticJsonConverter.cs:
```diff
@@ -7,7 +7,7 @@
 using Newtonsoft.Json.Linq;
 
-namespace Microsoft.VisualStudio.LanguageServices.Razor.Serialization
+namespace Microsoft.CodeAnalysis.Razor.Serialization
```
Moved to 1 shared location and removed the duplicated `Razor.LanguageServer.Common/Serialization/RazorDiagnosticJsonConverter.cs`

ajaybhargavb commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/JsonReaderExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+                }
+            }
+
+            throw new JsonSerializationException("Could not find string property '" + propertyName + "'.");
```
I know this was a cut-paste but for extra #pranavpoints:
```suggestion
            throw new JsonSerializationException($"Could not find string property '{propertyName}'.");
```

ajaybhargavb commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TagHelperResolutionResultComparer.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public bool Equals(TagHelperResolutionResult x, TagHelperResolutionResult y)
+        {
+            return x.Descriptors.SequenceEqual(y.Descriptors, TagHelperDescriptorComparer.Default) &&
```
Need to handle null case. It should be true if they are both null. Right now this would throw.

TanayParikh replied:
Added https://github.com/dotnet/aspnetcore-tooling/pull/2217/commits/56d1919dc743d9cc85b8b87d09220ce5bdf9b15f#diff-47a48387d941d7fe0445baa8f045e7c7R18-R26

ajaybhargavb commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/OOPTagHelperResolver.cs:
```diff
@@ -117,23 +117,16 @@ protected virtual async Task<TagHelperResolutionResult> ResolveTagHelpersOutOfPr
                 };
 
-                var resolutionJObject = await remoteClient.TryRunRemoteAsync<JObject>(
+                var resolutionJObject = await remoteClient.TryRunRemoteAsync<TagHelperResolutionResult>(
```
Rename? Because this is no longer a `JObject`.

NTaylorMullen commented on src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Performance/TagHelperResolutionResultSerializationBenchmark.cs:
```diff
@@ -0,0 +1,93 @@
+        private TagHelperResolutionResult TagHelperResolutionResult { get; }
+
+        [Benchmark(Description = "Razor TagHelperResolutionResult Roundtrip JObject Serialization")]
+        public void TagHelper_JObject_Serialization_RoundTrip()
```
It's good to capture this once to ensure we've improved the perf benchmark but I wouldn't keep it around. The other method represents how we want to deserialize everything though so definitely keep that around.

TanayParikh replied:
Removed!

NTaylorMullen commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/JsonReaderExtensions.cs:
```diff
@@ -53,5 +54,43 @@ public static string ReadNextStringProperty(this JsonReader reader, string prope
+                    results.Add(result);
+                }
+
+                if (reader.TokenType == JsonToken.EndObject)
```
Interesting, does the serializer.Deserialize not handle the EndObject token?

TanayParikh replied:
In a generic sense, I would expect it to. 

In practice, we don't in certain situations ([example](https://github.com/dotnet/aspnetcore-tooling/blob/06848c4725e55aff10972241801f69a184cca4c1/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs#L75-L77)). I played around with adding the end token reading there, but that led to some other issues, which made me suspect the omission may be by design? Just added this optional end token reading here as a safeguard.

davidfowl commented:
cc @JamesNK
